ddoc.d

```diff
@@ -0,0 +1,180 @@
+#!/usr/bin/env rdmd
+/**
+A wrapper around DDoc to allow custom extensions
+
+Copyright: D Language Foundation 2018
+
+License:   $(WEB boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Example usage:
+
+---
+./ddoc --compiler=<path-to-dmd> --output=out.html myfile.dd
+---
+
+Author: Sebastian Wilzbach
+*/
+import std.algorithm, std.array, std.ascii, std.conv, std.file, std.functional,
+        std.meta, std.path, std.range, std.string, std.typecons;
+import std.stdio : writeln, writefln;
+
+struct Config
+{
+    string dmdBinPath = "dmd";
+    string outputFile;
+}
+Config config;
+
+int main(string[] args)
+{
+    import std.getopt;
+    auto helpInformation = getopt(
+        args,
+        "compiler", "Compiler to use", &config.dmdBinPath,
+        "o|output", "Output file", &config.outputFile,
+    );
+
+    assert(config.outputFile, "An output file is required.");
+    assert(args.length > 1, "An input file is required.");
+
+    if (helpInformation.helpWanted)
+    {
+`DDoc wrapper
+./ddoc <file>...
+`.defaultGetoptPrinter(helpInformation.options);
+        return 1;
+    }
+
+    import std.file : readText;
+    auto text = args[$ - 1].readText;
+
+    // transform and extend the ddoc page
+    text = genHeader(text);
+
+    return compile(text, args[1 .. $ - 1]);
+}
+
+auto compile(R)(R buffer, string[] arguments)
+{
+    import std.process : pipeProcess, Redirect, wait;
+    auto args = [config.dmdBinPath, "-c", "-Df"~config.outputFile, "-o-"] ~ arguments;
+    args ~= "-";
+    auto pipes = pipeProcess(args, Redirect.stdin);
+    pipes.stdin.write(buffer);
+    pipes.stdin.close;
+    return wait(pipes.pid);
+}
+
+// replaces the content of a DDoc macro call
+auto updateDdocTag(string fileText, string ddocKey, string newContent)
+{
+    auto pos = fileText.representation.countUntil(ddocKey);
+    if (pos < 0)
+        return fileText;
+    const ddocStartLength = ddocKey.representation.until('(', No.openRight).count;
+    auto len = fileText[pos .. $].representation.drop(ddocStartLength).untilClosingParentheses.walkLength;
+    return fileText.replace(fileText[pos .. pos + len + ddocStartLength + 1], newContent);
+}
+
+// a range until the next ')', nested () are ignored
+auto untilClosingParentheses(R)(R rs)
```

Comment:
This reminded me of:
https://dlang.org/phobos-prerelease/std_algorithm_searching.html#.balancedParens

But that only returns a `bool`, perhaps another function could return a range of remaining elements from the last closing bracket, or empty if there isn't one.